test/roundrobin_test.go

```diff
@@ -45,36 +45,62 @@ const rrServiceConfig = `{"loadBalancingConfig": [{"round_robin":{}}]}`
+	// could land on backends owned by the previous LB policy.
 	backendCount := len(addrs)
-	for i := 0; i < backendCount; i++ {
-		for {
-			time.Sleep(time.Millisecond)
-			if ctx.Err() != nil {
-				return fmt.Errorf("timeout waiting for connection to %q to be up", addrs[i].Addr)
-			}
-			var peer peer.Peer
-			if _, err := client.EmptyCall(ctx, &testpb.Empty{}, grpc.Peer(&peer)); err != nil {
-				// Some tests remove backends and check if round robin is happening
-				// across the remaining backends. In such cases, RPCs can initially fail
-				// on the connection using the removed backend. Just keep retrying and
-				// eventually the connection using the removed backend will shutdown and
-				// will be removed.
-				continue
-			}
-			if peer.Addr.String() == addrs[i].Addr {
-				break
+	for j := 0; j < 2; j++ {
+		for i := 0; i < backendCount; i++ {
+			for {
+				time.Sleep(time.Millisecond)
+				if ctx.Err() != nil {
+					return fmt.Errorf("timeout waiting for connection to %q to be up", addrs[i].Addr)
+				}
+				var peer peer.Peer
+				if _, err := client.EmptyCall(ctx, &testpb.Empty{}, grpc.Peer(&peer)); err != nil {
+					// Some tests remove backends and check if round robin is happening
+					// across the remaining backends. In such cases, RPCs can initially fail
+					// on the connection using the removed backend. Just keep retrying and
+					// eventually the connection using the removed backend will shutdown and
+					// will be removed.
+					continue
+				}
+				if peer.Addr.String() == addrs[i].Addr {
+					break
+				}
 			}
 		}
 	}
-	// Make sure RPCs are sent to all backends.
-	for i := 0; i < 3*backendCount; i++ {
+
+	// At this point, we have seen RPCs being sent to all the backends that
+	// we care about. But this doesn't mean that the subConns in the
+	// round_robin picker are in the same order as that of addrs. So, we
+	// need to figure out the order of the subConns first.
+	firstIteration := make([]string, backendCount)
+	for i := 0; i < backendCount; i++ {
 		var peer peer.Peer
 		if _, err := client.EmptyCall(ctx, &testpb.Empty{}, grpc.Peer(&peer)); err != nil {
 			return fmt.Errorf("EmptyCall() = %v, want <nil>", err)
 		}
-		if gotPeer, wantPeer := peer.Addr.String(), addrs[i%backendCount].Addr; gotPeer != wantPeer {
-			return fmt.Errorf("rpc sent to peer %q, want peer %q", gotPeer, wantPeer)
+		firstIteration[i] = peer.Addr.String()
+	}
+
+	// Make another two passes and ensure the order is maintained.
+	var secondIteration []string
+	for i := 0; i < 2; i++ {
+		for _, want := range firstIteration {
+			var peer peer.Peer
+			if _, err := client.EmptyCall(ctx, &testpb.Empty{}, grpc.Peer(&peer)); err != nil {
+				return fmt.Errorf("EmptyCall() = %v, want <nil>", err)
+			}
+			got := peer.Addr.String()
+			secondIteration = append(secondIteration, got)
+			if got != want {
+				return fmt.Errorf("non-roundrobin, first iter: %v, second iter: %v", firstIteration, secondIteration)
```

Comment:
Can we wait until the end to check & fail so we can show all of `secondIteration` instead of the partial second information up to the point where the deviation occurred?

Or maybe something like:

```go
// Perform 3 iterations and ensure they are all equal
var iterations [][]string
for i := 0; i < 3; i++ {
	iteration := make([]string, backendCount)
	for c := 0; c < backendCount; c++ {
		var peer peer.Peer
		if _, err := client.EmptyCall(ctx, &testpb.Empty{}, grpc.Peer(&peer)); err != nil {
			return fmt.Errorf("EmptyCall() = %v, want <nil>", err)
		}
		iteration[c] = peer.Addr.String()
	}
	iterations = append(iterations, iteration)
}
// check iterations[0] == iterations[1] == iterations[2] and log all 3 if they don't match
```

Reply:
Done. Thanks.